### PR TITLE
Add network requests allowed API

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
Things to consider before you submit the PR:

* [X] Has `CHANGELOG.md` been updated?
* [X] Are tests passing locally?
* [X] Are the files formatted correctly?
* [X] Did you add unit tests if this modifies the Windows code?
* [X] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [ ] Has at least one reviewer tested the PR?

## Description

Add API to block network requests.

## Related PRs or issues

[AB#85219](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/85219)
[AB#85223](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/85223)

### CODE REVIEW

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/IAppCenterService.cs [hunk 1]
```diff
@@ -20,6 +20,11 @@ public interface IAppCenterService
         /// </summary>
         bool InstanceEnabled { get; set; }
 
+        /// <summary>
+        /// Check whether network requests are allowed or disallowed.
+        /// </summary>
+        bool IsNetworkRequestsAllowed { get; set; }
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/IAppCenterService.cs:26]: Remove it

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: Apps/Contoso.Android.Puppet/Constants.cs [hunk 2]
```diff
@@ -11,5 +11,6 @@ public static class Constants
         public const string Warning = "Warning";
         public const string Error = "Error";
         public const string StorageSizeKey = "StorageSizeKey";
+        public const string AllowNetworkRequests = "allow-network-requests";
```
Comment (MatkovIvan) [inline Apps/Contoso.Android.Puppet/Constants.cs:14]: Let's make it consistent 
```suggestion
        public const string AllowNetworkRequestsKey = "AllowNetworkRequests";
```

File: Apps/Contoso.Android.Puppet/ModulePages/AppCenterFragment.cs [hunk 3]
```diff
@@ -117,6 +123,15 @@ private async void UpdateEnabled(object sender, CompoundButton.CheckedChangeEven
             AppCenterEnabledSwitch.Checked = await AppCenter.IsEnabledAsync();
         }
 
+        private void NetworkRequestAllowedChange(object sender, CompoundButton.CheckedChangeEventArgs e)
+        {
+            AppCenter.IsNetworkRequestsAllowed = e.IsChecked;
+            var prefs = Context.GetSharedPreferences("AppCenter", FileCreationMode.Private);
```
Comment (MatkovIvan) [inline Apps/Contoso.Android.Puppet/ModulePages/AppCenterFragment.cs:129]: Move(or reuse if exists) it to consts

File: Apps/Contoso.Android.Puppet/Resources/values/Strings.xml [hunk 4]
```diff
@@ -13,6 +13,7 @@
     <string name="AppCenterWriteUserId">Write User Id</string>
     <string name="AppCenterUserId">User Id</string>
     <string name="AnalyticsEnabled">Analytics Enabled</string>
+    <string name="AppcenterNetworkRequestsAllowed">Allow network requests</string>
```
Comment (MatkovIvan) [inline Apps/Contoso.Android.Puppet/Resources/values/Strings.xml:16]:
```suggestion
    <string name="AppCenterNetworkRequestsAllowed">Allow network requests</string>
```

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/Constants.cs [hunk 5]
```diff
@@ -8,6 +8,7 @@ public static class Constants
     {
         public const string BingMapsAuthKey = "paste-key-here";
         public const string UserId = "userId";
+        public const string AllowNetworkRequests = "allow-network-requests";
```
Comment (MatkovIvan) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/Constants.cs:11]:
```suggestion
        public const string AllowNetworkRequestsKey = "AllowNetworkRequests";
```

File: Apps/Contoso.iOS.Puppet/Constants.cs [hunk 6]
```diff
@@ -11,5 +11,6 @@ public static class Constants
         public const string Warning = "Warning";
         public const string Error = "Error";
         public const string StorageSizeKey = "AppCenterStorageSizeKey";
+        public const string AllowNetworkRequests = "allow-network-requests";
```
Comment (MatkovIvan) [inline Apps/Contoso.iOS.Puppet/Constants.cs:14]: Same

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelGroup.cs [hunk 7]
```diff
@@ -26,6 +26,8 @@ public sealed class ChannelGroup : IChannelGroup, IAppSecretHolder
 
         private bool _isShutdown;
 
+        private bool _networkRequestsAllowed = true;
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelGroup.cs:29]: Unused

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/IngestionDecorator.cs [hunk 8]
```diff
@@ -35,5 +35,15 @@ public virtual void Dispose()
         {
             DecoratedApi.Dispose();
         }
+
+        public void SetEnabled(bool enabled, bool deleteLogs)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/Http/IngestionDecorator.cs:39]: We do not need second bool in `Ingestion`. Let's make it as a property.

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/IIngestion.cs [hunk 9]
```diff
@@ -18,6 +18,19 @@ public interface IIngestion : IDisposable
...
+        /// </summary>
+        /// <param name="enabled">Value indicating whether channel should be enabled or disabled</param>
+        /// <param name="deleteLogs">True if logs should be deleted, false otherwise.</param>
+        void SetEnabled(bool enabled, bool deleteLogs);
+
+        /// <summary>
+        /// Gets value indicating whether the ingestion is enabled.
+        /// </summary>
+        /// <returns>True if enabled, false otherwise.</returns>
+        bool IsEnabled();
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Ingestion/IIngestion.cs:32]:
```suggestion
        /// <summary>
        /// The value indicating whether the ingestion is enabled.
        /// </summary>
        bool Enabled { get; set; }
```

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs [hunk 10]
```diff
@@ -50,6 +50,10 @@ protected override void OnStart()
                 AppCenter.LogLevel = LogLevel.Verbose;
                 AppCenterLog.Info(LogTag, "AppCenter.LogLevel=" + AppCenter.LogLevel);
                 AppCenterLog.Info(LogTag, "AppCenter.Configured=" + AppCenter.Configured);
+                if (Current.Properties.ContainsKey(Constants.AllowNetworkRequests) && Current.Properties[Constants.AllowNetworkRequests] is bool isAllowedNetworkRequest)
```
Comment (DmitriyKirakosyan) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs:53]:
```suggestion
                if (Current.Properties.ContainsKey(Constants.AllowNetworkRequests) && Current.Properties[Constants.AllowNetworkRequests] is bool isNetworkRequestsAllowed)
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs [hunk 11]
```diff
@@ -86,6 +89,28 @@ static LogLevel PlatformLogLevel
...
+        /// Allow or disallow network requests. True by the default.
+        /// </summary>
+        public static bool PlatformIsNetworkRequestsAllowed
+        {
+            get => Instance._networkRequestsAllowed;
+            set 
+            {
+                if (Instance._networkRequestsAllowed == value)
+                {
+                    AppCenterLog.Info(AppCenterLog.LogTag, $"Network request already {(value ? "allowed" : "disallowed")}");
```
Comment (DmitriyKirakosyan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs:102]:
```suggestion
                    AppCenterLog.Info(AppCenterLog.LogTag, $"Network requests are already {(value ? "allowed" : "disallowed")}");
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs [hunk 12]
```diff
@@ -86,6 +89,28 @@ static LogLevel PlatformLogLevel
...
+                    return;
+                }
+                Instance._networkRequestsAllowed = value;
+                if (Instance._channelGroup != null)
+                {
+                    Instance._channelGroup.SetEnabled(Instance.InstanceEnabled && value, false);
+                }
+                AppCenterLog.Info(AppCenterLog.LogTag, $"Set network request allowed {Instance._networkRequestsAllowed}");
```
Comment (DmitriyKirakosyan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs:110]:
```suggestion
                AppCenterLog.Info(AppCenterLog.LogTag, $"Set network requests {(Instance._networkRequestsAllowed ? "allowed" : "disallowed")}");
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs [hunk 13]
```diff
@@ -86,6 +87,35 @@ static LogLevel PlatformLogLevel
             set => AppCenterLog.Level = value;
         }
 
+        /// <summary>
+        /// Allow or disallow network requests. True by the default.
+        /// </summary>
+        public static bool PlatformIsNetworkRequestsAllowed
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/AppCenter.cs:93]: It looks like it's not thread safe. Recheck and add `AppCenterLock` if needed

File: Apps/Contoso.WPF.Puppet.DotNetCore/App.xaml.cs [hunk 14]
```diff
@@ -45,6 +45,7 @@ protected override void OnStartup(StartupEventArgs e)
             }
 
             // Start AppCenter.
+            AppCenter.IsNetworkRequestsAllowed = AppCenter.IsNetworkRequestsAllowed;
```
Comment (Jamminroot) [inline Apps/Contoso.WPF.Puppet.DotNetCore/App.xaml.cs:48]: I might be missing something here due to lack of context highlight. Can you explain this line?

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelGroup.cs [hunk 15]
```diff
@@ -114,6 +114,14 @@ public void SetEnabled(bool enabled)
             }
         }
 
+        public void SetNetworkRequest(bool isAllowed)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/ChannelGroup.cs:117]:
```suggestion
        public void SetNetworkRequestAllowed(bool isAllowed)
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 16]
```diff
@@ -555,14 +555,30 @@ private void HandleSendingFailure(State state, string batchId, Exception excepti
             }
         }
 
+        public void SetNetworkRequest(bool isAllowed)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs:558]:
```suggestion
        public void SetNetworkRequestAllowed(bool isAllowed)
```

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 17]
```diff
@@ -555,14 +555,30 @@ private void HandleSendingFailure(State state, string batchId, Exception excepti
...
+        {
+            if (isAllowed)
+            {
+                Resume(_mutex.State);
+            }
+            else
+            {
+                Suspend(_mutex.State, false, new CancellationException());
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs:566]: It disables the channel and it will be re-enabled by service.

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 18]
```diff
@@ -282,7 +284,6 @@ private void Resume(State state)
             {
                 using (_mutex.GetLock(state))
                 {
-                    _enabled = true;
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs:285]: It MUST stay under the lock/mutex

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs [hunk 19]
```diff
@@ -275,14 +275,17 @@ public async Task ClearAsync()
             }
         }
 
-        private void Resume(State state)
+        private void Resume(State state, bool needEnableChannel = true)
```
Comment (MatkovIvan) [inline SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Channel/Channel.cs:278]: This _should_ work, but please add some comments about such parameters

**Review by MatkovIvan — APPROVED** [review]

**Pull request merged**